torch/csrc/jit/tensorexpr/external_functions.cpp

```diff
@@ -196,6 +204,29 @@ void nnc_aten_mean(
   }
 }
 
+void nnc_aten_addmm(
```

Comment:
If we add a new external call function, let's add a test for it as well (they live in https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_external_calls.cpp).

Another option is to not include `addmm` in this PR at all - for this particular op it might make sense to decompose it to `add` + `mm`, both of which are already present.

Reply:
> If we add a new external call function, let's add a test for it as well (they live in https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_external_calls.cpp).

Test case added.

> Another option is to not include `addmm` in this PR at all - for this particular op it might make sense to decompose it to `add` + `mm`, both of which are already present.

@ZolotukhinM I feel this kinda of decomposition/translation would introduce complexity (and slightly more overhead). I prefer we either express an operator with pure TensorExpression or with straightforward extern function call. The operator schema in the TorchScript IR and in the ATen/Functions.h interface derive from the same source (native_functions.yaml). I think we should simply stick to the exact mapping and (eventually) automatically generate the boilerplate code.

Reply:
I personally slightly prefer the other way (mainly because we reuse code rather than adding a ver similar piece), but I'm fine with either one. I also see your point about sticking to the 1-1 mapping, and it makes perfect sense to me as well. Btw, @Chillee was interested in looking into generating the ext-calls wrappers automatically too, so maybe we could start a discussion about it.

Reply:
P.S. just checked the ATen native cpu implementation, seems "matmul" calls "mm", which in turn calls "addmm", which then calls gemm implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L1094

So "addmm" this more "basic" in this regards, lol.

Reply:
> if we can directly invoke QNNPack / XNNPack micro-kernels from NNC-generated assembly.

I am not sure how straight forward this is going to be. Mainly because there is a state to be created before. In TVM based approaches where equivalent would have been to actually write weight transformation op and then conv/gemm op. Then weight transformation op can be "precomputed" based on const propagation.

In this case I imagine we would have to have two microkernels. 1) weight prepacking 2) actual op. Then presume weight prepacking can be const propagated.

However, there are a few challenges with these approaches:
1. It introduces a tight coupling between NNC and QNNPACK/XNNPACK. QNNPACK fork being internal it might be relatively easy in practice but not in good in theory. The implementations of internals of QNNPACK op have now been exposed to outside and going forward we need to keep them in sync. For example if for particular kernel that produces and output tile of 8x8 for gemm but changes the packing of weights, will now require coordination with NNC. Doing this with XNNPACK will be even harder.
2.  For equivalent perf we need to have threading enabled the same way in XNNPACK/QNNPACK since NNC runtime will invoke ukernels differently compared to NNC. This is where efficiency of threadpool might matter too.

I think the easiest path for invoking these ops is based off of their prepacked variant/quantized variant. Like `prepacked::linear_pack`, `prepacked::linear_run` or `quantized::linear_prepack`, `quantized::linear_run`.